### PR TITLE
Integrating Task2 & task 7

### PR DESCRIPTION
**Pull request by fennievdg**
Built the unicorn_eval docker container and ran evaluate.py for both Task 2 and Task 7, without errors. 

Here is a list of explanations for changes mainly in evaluation.py and utils.py. 
Dockerfile:
Change: Space related error when building docker, so no cache-dir when installing packages added 

Evaluation.py
Changes:
	- def process(job), image origin, image direction needed to be extracted
	- def get_cases_extra_labels_detection expected 2d coordinates, but in radiology tasks we have 3d coordinates, so currently it just selects the first 2 coordinates, does not really matter since radiology tasks (as far as I know) do not use the pixel spacing. 
	- I needed the whole 'cases_extra_labels' returned so I could later extract diameter from it for the metric for task 7. 

Utils.py
 Changes: 
	- def adapt_features: test and train image spacing, origins, directons added, and right now I assumed I am the only MLP__ for detection 
	- def evaluate_predictions: For some reason when I looked at my predictions file for task 7 it only contained one prediction per caseid; it seemed like it was trying to match the number of caseids with the number of predictions. However I do want all of my nodule coordinates (several per caseid) to get saved in the predictions.json so I adapted how predictions gets appended to metrics list for task 7. 
	- def extract_points: Issue was it did not account for when a gt has 3 coordinates, which is the case for task 7. So I updated this to pull out the first 2 coordinates when only x, y are present, and if x,y,z are present it pulls these out. 
	- For task 7 I wanted case_extra_labels to be a particular format for the diameter which includes the coordinates (points), the diameter, and the caseid 
	- def extract_embeddings_and_labels: added image origin, image direction

Thanks for the review!

### CODE REVIEW

**Review by clemsgrs — CHANGES_REQUESTED** [review]
Left comments to improve code readability.
The main concern i have is about the adaptor for task 2: the idea of the eval repo is to have code **as generic as possible**, such that i can easily be applied to a new dataset / new task.

The `LinearClassifierAdaptor` you proposed has some flaws:
- the name doesn't reflect what the adaptor is / does
- the code cannot be applied on a new classification task

Still, there were some nice changes i've integrated into a refactored `MultiLayerPerceptron` class, which handles classification tasks using a MLP and some logic to allow for early stopping. This class should be generic enough that i can be re-used for task 2 as well.

File: src/unicorn_eval/metrics/auc.py [hunk 1]
```diff
@@ -12,27 +12,73 @@
...
 from sklearn.metrics import roc_auc_score
 
 
-class ModelEvaluator:
-    def __init__(self, predictions, ground_truths):
-        # predictions: list of tuples (scan_id, predicted_value)
-        # ground_truths: dict {scan_id: true_label}
-        self.predictions = predictions
-        self.ground_truths = ground_truths
+class RocAuc:
```
Comment (clemsgrs) [inline src/unicorn_eval/metrics/auc.py:19]: if we stick to the `MultiLayerPerceptron` adaptor for task 2, we get two arrays:
- `predictions` which contains predictions (or probabilities) for each test case
- `case_labels` which contains groundtruth labels for each test case 

depending on the task, we might have an extra `case_extra_labels`

computing auc should be as simple as running `roc_auc_score(case_labels, predictions)`

File: src/unicorn_eval/adaptors/detection.py [hunk 2]
```diff
@@ -356,4 +356,252 @@ def predict(self) -> list:
             patch_size=self.patch_size,
         )
 
-        return predicted_points
\ No newline at end of file
+        return predicted_points
+    
+class MLPRegressor(nn.Module, PatchLevelTaskAdaptor):
```
[the remark this answers is not in the document]
Comment (clemsgrs) [reply src/unicorn_eval/adaptors/detection.py:361]: fixed

File: src/unicorn_eval/metrics/sensitivity.py [hunk 3]
```diff
@@ -0,0 +1,337 @@
...
+    Convert run‑time tensors/lists to the CSVs required by the unchanged
+    CPM evaluator and return the metric value.
+    """
+    with tempfile.TemporaryDirectory() as tmp_dir_str:
+        tmp_dir = Path(tmp_dir_str)
+
+        # ---- seriesuids.csv ------------------------------------------------
+        series_csv = _dump([[cid] for cid in case_ids], tmp_dir,
+                           "seriesuids.csv")
+        print("DEBUG  seriesuids    :", case_ids)
```
[the remark this answers is not in the document]
Comment (fennievdg) [reply src/unicorn_eval/metrics/sensitivity.py:299]: removed

File: src/unicorn_eval/metrics/sensitivity.py [hunk 4]
```diff
@@ -0,0 +1,337 @@
...
+            case_ids, test_labels, test_extra_labels
+        ):
+            diam_list = diam_struct[0]["points"]
+            for (x, y, z), diam_rec in zip(coords_case, diam_list):
+                ann_rows.append([
+                    cid, f"{x}", f"{y}", f"{z}",
+                    f"{float(diam_rec['diameter'])}"
+                ])
+        ann_csv = _dump(ann_rows, tmp_dir, "annotations.csv")
+        print("DEBUG  ann_rows (GT) :", len(ann_rows) - 1, "rows")
```
[the remark this answers is not in the document]
Comment (fennievdg) [reply src/unicorn_eval/metrics/sensitivity.py:314]: removed

File: src/unicorn_eval/metrics/sensitivity.py [hunk 5]
```diff
@@ -0,0 +1,337 @@
...
+                      coordX_label, coordY_label, coordZ_label,
+                      CADProbability_label]]
+        for row in test_predictions:
+            test_id, x, y, z, p = row
+            if test_id in case_ids:          # ignore stray predictions
+                cand_rows.append([
+                    test_id, f"{x}", f"{y}", f"{z}", f"{p}"
+                ])
+        cand_csv = _dump(cand_rows, tmp_dir, "candidates.csv")
+        print("DEBUG  cand_rows (CAD):", len(cand_rows) - 1, "rows")
```
[the remark this answers is not in the document]
Comment (clemsgrs) [reply src/unicorn_eval/metrics/sensitivity.py]: removed it

File: src/unicorn_eval/evaluate.py [hunk 6]
```diff
@@ -92,15 +105,26 @@
...
     """Processes a single algorithm job, looking at the outputs"""
+
+    image_origin, image_direction = None, None
+    embeddings = coordinates = spacing = patch_size = None
+    image_size = image_spacing = prediction = None
     report = "Processing:\n"
     report += pformat(job)
     report += "\n"
 
+    print(f"\n===== JOB {job['pk']} =====", flush=True)
```
[the remark this answers is not in the document]
Comment (clemsgrs) [reply src/unicorn_eval/evaluate.py:195]: removed it

File: src/unicorn_eval/evaluate.py [hunk 7]
```diff
@@ -92,15 +105,26 @@
...
+    mapping_path = GROUNDTRUTH_DIRECTORY / "mapping.csv"
+    print("mapping.csv:", mapping_path, flush=True)
```
[the remark this answers is not in the document]
Comment (clemsgrs) [reply src/unicorn_eval/evaluate.py:197]: removed it

File: src/unicorn_eval/evaluate.py [hunk 8]
```diff
@@ -399,6 +440,12 @@ def main():
                 patch_size=patch_size,
                 test_image_sizes=case_image_size,
                 shot_extra_labels=shot_extra_labels,
+                test_image_spacing=case_image_spacings,
+                test_image_origins=case_image_origins, 
+                test_image_directions=case_image_directions,
+                train_image_spacing=shot_image_spacings,
+                train_image_origins=shot_image_origins,
```
[the remark this answers is not in the document]
Comment (clemsgrs) [reply src/unicorn_eval/evaluate.py]: replaced it

File: src/unicorn_eval/evaluate.py [hunk 9]
```diff
@@ -399,6 +440,12 @@ def main():
                 patch_size=patch_size,
                 test_image_sizes=case_image_size,
                 shot_extra_labels=shot_extra_labels,
+                test_image_spacing=case_image_spacings,
+                test_image_origins=case_image_origins, 
+                test_image_directions=case_image_directions,
+                train_image_spacing=shot_image_spacings,
+                train_image_origins=shot_image_origins, 
+                train_image_directions=shot_image_directions,
```
[the remark this answers is not in the document]
Comment (clemsgrs) [reply src/unicorn_eval/evaluate.py]: replaced it

File: src/unicorn_eval/utils.py [hunk 10]
```diff
@@ -89,7 +103,14 @@ def adapt_features(
     patch_size=224,
     test_image_sizes=None,
     shot_extra_labels=None,
+    test_image_spacing=None,
+    test_image_origins=None,
+    test_image_directions=None,
+    train_image_spacing=None,
```
[the remark this answers is not in the document]
Comment (clemsgrs) [reply src/unicorn_eval/utils.py]: replaced it

File: src/unicorn_eval/utils.py [hunk 11]
```diff
@@ -89,7 +103,14 @@ def adapt_features(
     patch_size=224,
     test_image_sizes=None,
     shot_extra_labels=None,
+    test_image_spacing=None,
+    test_image_origins=None,
+    test_image_directions=None,
+    train_image_spacing=None,
+    train_image_origins=None,
```
[the remark this answers is not in the document]
Comment (clemsgrs) [reply src/unicorn_eval/utils.py]: replaced it

**Review by clemsgrs — APPROVED** [review]
